### PR TITLE
Preliminary support for open-anything-or-use-file functions

### PR DESCRIPTION
**Pull request by gonzalocasas**
Add functions to open and iterate over files with the goal of accepting filename/path-like objects or file-like objects everywhere in COMPAS. 

I have replaced most of the occurrences of methods accepting only filepaths to use this new thingy. The one exception is GLTF because there's some detection of type based on the filepath that makes it a bit tricky and I'd need to discuss with @beverlylytle  first.

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/compas/files/test_iotools.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+def test_iter_file_chunks_path_image(path_image):
+    CHUNK_SIZE = 30
+    chunks = []
+
+    with _iotools.open_file(path_image, 'rb') as file:
+        for data in _iotools.iter_file(file, size=CHUNK_SIZE):
+            chunks.append(data)
+
+    assert len(chunks) == math.ceil(IMAGE_FILE_SIZE / float(CHUNK_SIZE))
```
Comment (beverlylytle) [inline tests/compas/files/test_iotools.py]: What about tests for mode='w'?

File: src/compas/files/_iotools.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+@contextmanager
+def open_file(file_or_filename, mode='r'):
+    """Context-manager to open a file, path or URL and return a corresponding file object.
+
+    If the argument is a path-like object, it will open it,
+    if the argument is an HTTP(s) URL, it will open it, and
+    if the argument is already a file-like object, it will simply yield it.
+
+    This context manager ensures that when using a path-like object as input,
+    the file is closed once it it goes out of scope.
```
Comment (beverlylytle) [inline src/compas/files/_iotools.py]: I think the fact that a file-like object is not closed by this context manager should really be underscored in the docs.

File: tests/compas/files/test_iotools.py (hunk 1 above)
Comment (gonzalocasas) [reply tests/compas/files/test_iotools.py]: good point! added on https://github.com/compas-dev/compas/pull/712/commits/623b58d9bd47ffa9ef306ff4f486b9acaad02eca

File: src/compas/files/_iotools.py (hunk 2 above)
Comment (gonzalocasas) [reply src/compas/files/_iotools.py]: added a note about that on https://github.com/compas-dev/compas/pull/712/commits/0f707368d373675e0c6b7fcf646c7f5812038d5d

File: tests/compas/files/test_iotools.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+def test_open_file_write_file_object():
+    path = os.path.join(tempfile.gettempdir(), 'test-file.txt')
+    with open(path, mode='w') as f:
+        with _iotools.open_file(f) as file:
+            file.write('Hello world')
+
+    with open(path, mode='r') as f:
+        with _iotools.open_file(f) as file:
+            assert file.read() == 'Hello world'
```
Comment (beverlylytle) [inline tests/compas/files/test_iotools.py]: What about a test that checks that a file-like object is not closed after exiting the `with`? Edit: or that things are closed when they are expected to be...

Comment (gonzalocasas) [reply tests/compas/files/test_iotools.py]: added a couple of tests to check for close/open conditions: https://github.com/compas-dev/compas/pull/712/commits/126ee212df61601a904fde965fdf8e991608648c

**Review by beverlylytle — APPROVED** [review]
LGTM!

**Comment by gonzalocasas** [comment]
fyi: not merging just now because it's a new feature and we're in feature-freeze before 1.0 release. In the meantime, I will try to extend this change to other parts of `compas.files` and will change this to non-draft early next year after 1.0 is already out.

File: tests/compas/test_iotools.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+
+@pytest.fixture
+def url_text():
+    return 'https://raw.githubusercontent.com/compas-dev/compas/master/README.md'
+
+
+@pytest.fixture
+def url_image():
+    return 'http://google.com/favicon.ico'
```
Comment (beverlylytle) [inline tests/compas/test_iotools.py]: Is this legal?  Maybe to be on the safe side choose a clearly open source image.

File: tests/compas/test_iotools.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def test_open_file_object_text(path_text):
+    with open(path_text, mode='r') as f:
+        with _iotools.open_file(f) as file:
+            assert len(file.read()) == TEXT_FILE_SIZE
+
+
+def test_open_file_memory_stream():
+    text = b"How many roads must a man walk down?"
```
Comment (beverlylytle) [inline tests/compas/test_iotools.py]: Again, legality?  Especially since this was just sold....

Comment (gonzalocasas) [reply tests/compas/test_iotools.py]: hehe..! changed to public domain and CreativeCommons image! ;) https://github.com/compas-dev/compas/pull/712/commits/7b15f1677a5c6f540b5a5b78ccab8c395ee03955

**Review by tomvanmele — APPROVED** [review]
LGTM...

perhaps some stuff i noticed for discussion during the meeting

* move from/to json to base
* create `compas.data` for base and decoder/encoder
* remove geometry duplicates from `compas.robots`

**Pull request merged**